File: python-app/application/strategies/abstracts/base_application_strategy.py
from abc import ABC
from domain.abstracts.base_domain import BaseDomain
from infrastructure.repository.repository import Repository

class BaseApplicationStrategy(ABC):
    def __init__(self, domain):
        self.repo = Repository()
        self.domain_class = domain

    def ensure_domain(self) -> BaseDomain:
        return self.repo.build_domain(self.domain_class)

    def list_domains(self):
        models = self.ensure_domain().list_models()
        header = f"=== {self.domain_class.get_domain_name()} Models ==="
        return header, models, ""

    def execute(self, prompt: str) -> str:
        return self.ensure_domain().send_message(prompt)

File: angular-app/src/core/application/strategies/abstracts/base-application-strategy.ts
import { BaseDomain } from '../../../domain/abstracts/base-domain';
import { Repository } from '../../../infrastructure/repository';
import { IServer } from '../../../infrastructure/servers';

export interface DomainConstructor<T extends BaseDomain = BaseDomain> {
  new (
    server: IServer,
    modelName: string
  ): T;
  name: string;
  getDomainName(): string;
}

export interface DomainListView {
  header: string;
  names: string[];
  prefix: string;
}

export abstract class BaseApplicationStrategy {
  private domainInstance: BaseDomain | null = null;

  protected constructor(
    public readonly domainClass: DomainConstructor,
    private readonly repository: Repository = new Repository()
  ) {}

  protected ensureDomain(): BaseDomain {
    if (!this.domainInstance) {
      this.domainInstance = this.repository.buildDomain(this.domainClass);
    }
    return this.domainInstance;
  }

  useModel(model: string | undefined): void {
    this.ensureDomain().useModel(model);
  }

  getCurrentModel(): string {
    return this.ensureDomain().model;
  }

  async execute(task: string): Promise<string> {
    return this.ensureDomain().sendMessage(task);
  }

  async listDomains(): Promise<DomainListView> {
    const names = await this.ensureDomain().listModels();
    return {
      header: `=== ${this.domainClass.getDomainName()} Models ===`,
      names,
      prefix: '',
    };
  }
}
